### PR TITLE
Answer controller delete answer

### PR DESCRIPTION
**Pull request by prateeksingh**
Added endpoint api to delete answer. please review the code.

### CODE REVIEW

File: quora-service/src/main/java/com/upgrad/quora/service/entity/AnswerEntity.java [hunk 1]
```diff
@@ -7,6 +7,12 @@
 
 @Entity
 @Table(name = "answer")
+@NamedQueries(
+        {
+                @NamedQuery(name = "answerEntityByQuestionId", query = "select i from QuestionEntity i where i.question_id = :uuid"),
+                @NamedQuery(name = "answerEntityById", query = "select i from QuestionEntity i where i.id = :id")
```
Comment (parimalvfx) [inline quora-service/src/main/java/com/upgrad/quora/service/entity/AnswerEntity.java]:
```suggestion
                @NamedQuery(name = "answerEntityById", query = "select i from AnswerEntity i where i.id = :id")
```

File: quora-service/src/main/java/com/upgrad/quora/service/dao/AnswerDao.java [hunk 2]
```diff
@@ -1,19 +1,47 @@
...
+        try {
+            return entityManager.createNamedQuery("questionEntityByUuid", AnswerEntity.class).setParameter("uuid",uuid).getSingleResult();
+        } catch (NoResultException nre) {
+            return null;
+        }
+    }
+
+    public void userQuestionDelete(String uuid) {
+        answerEntity = getAnswerByUuid(uuid);
+        entityManager.remove(answerEntity);
```
Comment (parimalvfx) [inline quora-service/src/main/java/com/upgrad/quora/service/dao/AnswerDao.java]: Have you implemented `DELETE CASCADE`?

File: quora-service/src/main/java/com/upgrad/quora/service/entity/AnswerEntity.java [hunk 3]
```diff
@@ -7,6 +7,12 @@
 
 @Entity
 @Table(name = "answer")
+@NamedQueries(
+        {
+                @NamedQuery(name = "answerEntityByQuestionId", query = "select i from QuestionEntity i where i.question_id = :uuid"),
```
Comment (parimalvfx) [inline quora-service/src/main/java/com/upgrad/quora/service/entity/AnswerEntity.java]:
```suggestion
                @NamedQuery(name = "answerEntityByQuestionId", query = "select i from AnswerEntity i where i.question_id = :id"),
```

File: quora-service/src/main/java/com/upgrad/quora/service/business/AnswerBusinessService.java [hunk 4]
```diff
@@ -48,4 +50,41 @@ public AnswerEntity createAnswer(final AnswerEntity answerEntity, final String q
         answerEntity.setQuestion(questionEntity);
         return answerDao.createAnswer(answerEntity);
     }
+
```
Comment (parimalvfx) [inline quora-service/src/main/java/com/upgrad/quora/service/business/AnswerBusinessService.java]: Add this line above deleteAnswer method:

```suggestion
@Transactional(propagation = Propagation.REQUIRED)
```

File: quora-service/src/main/java/com/upgrad/quora/service/business/AnswerBusinessService.java [hunk 5]
```diff
@@ -48,4 +50,41 @@ public AnswerEntity createAnswer(final AnswerEntity answerEntity, final String q
...
+        }
+        if(userAuthEntity.getUser().getRole().equals("nonadmin") || !userAuthEntity.getUser().equals(answerDao.getAnswerByUuid(uuid).getUser())){
+            throw new AuthorizationFailedException("ATHR-003", "Only the answer owner or admin can delete the answer");
+
+        }
+        if (answerDao.getAnswerByUuid(uuid)==null) {
+            throw new AnswerNotFoundException("AND-001","Entered answer uuid does not exist.");
+        }
+
+        answerDao.userQuestionDelete(uuid);
```
Comment (parimalvfx) [inline quora-service/src/main/java/com/upgrad/quora/service/business/AnswerBusinessService.java]: Internally we should delete answer by `id` and not `uuid`.

File: quora-api/src/main/java/com/upgrad/quora/api/controller/AnswerController.java [hunk 6]
```diff
@@ -28,4 +32,28 @@
         AnswerResponse answerResponse = new AnswerResponse().id(createdAnswerEntity.getUuid()).status("ANSWER CREATED");
         return new ResponseEntity<AnswerResponse>(answerResponse, HttpStatus.CREATED);
     }
+
+    @RequestMapping(method = RequestMethod.POST, path = "/answer/delete/{answerId}", consumes = MediaType.APPLICATION_JSON_UTF8_VALUE, produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
```
Comment (parimalvfx) [inline quora-api/src/main/java/com/upgrad/quora/api/controller/AnswerController.java]: It should be a DELETE request:

```suggestion
    @RequestMapping(method = RequestMethod.DELETE, path = "/answer/delete/{answerId}", consumes = MediaType.APPLICATION_JSON_UTF8_VALUE, produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
```

File: quora-api/src/main/java/com/upgrad/quora/api/controller/AnswerController.java [hunk 7]
```diff
@@ -28,4 +32,28 @@
...
+    public ResponseEntity<AnswerResponse> deleteAnswer(@PathVariable("questionId") final String questionId, @RequestHeader("authorization") final String authorization) throws AuthorizationFailedException, AnswerNotFoundException {
```
Comment (parimalvfx) [inline quora-api/src/main/java/com/upgrad/quora/api/controller/AnswerController.java]: We should use `AnswerDeleteResponse` and not `AnswerResponse`:

```suggestion
    public ResponseEntity<AnswerDeleteResponse> deleteAnswer(@PathVariable("questionId") final String questionId, @RequestHeader("authorization") final String authorization) throws AuthorizationFailedException, AnswerNotFoundException {
```

File: quora-service/src/main/java/com/upgrad/quora/service/dao/AnswerDao.java (hunk 2 above)
Comment (prateeksingh) [reply quora-service/src/main/java/com/upgrad/quora/service/dao/AnswerDao.java]: No, but as per requirement api should delete answer only. Also Delete Cascade would delete all child and Answer is the last node.

**Pull request merged**